src/utils/database-common/src/transactions/db_transaction_manager.rs

```diff
@@ -0,0 +1,188 @@
+    let db_transaction_manager = base_catalog
+        .get_one::<dyn DatabaseTransactionManager>()
+        .unwrap();
+
+    // Start transaction
+    let transaction_ref = db_transaction_manager.make_lazy_transaction_ref().await?;
+
+    // Create a chained catalog for transaction-aware components,
+    // but keep a local copy of transaction pointer
+    let chained_catalog = CatalogBuilder::new_chained(base_catalog)
+        .add_value(transaction_ref.clone())
+        .build();
+
+    // Run transactional code in the callback
+    let result = callback(chained_catalog).await;
+
+    // Commit or rollback transaction depending on the result
+    match result {
+        // In case everything succeeded, commit the transaction
+        Ok(_) => {
+            db_transaction_manager
+                .commit_transaction(transaction_ref)
+                .await?;
+            Ok(())
+        }
+
+        // Otherwise, do an explicit rollback
+        Err(e) => {
+            db_transaction_manager
+                .rollback_transaction(transaction_ref)
+                .await?;
+            Err(e)
+        }
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+/// Represents a shared reference to [`sqlx::Transaction`] that unifies how we
+/// propagate transactions across different database implementations. This
+/// reference can appear in multiple components simultaneously, but can be used
+/// from only one place at a time via locking. Despite its async nature, this
+/// lock must be held only for the duration of the DB query and released when
+/// passing control into other components.
+#[derive(Clone)]
+pub struct LazyTransactionRef {
+    inner: Arc<Mutex<LazyTransactionRefInner>>,
+}
+
+impl LazyTransactionRef {
+    pub fn new<DB: sqlx::Database>(connection_pool: sqlx::pool::Pool<DB>) -> Self {
+        Self {
+            inner: Arc::new(Mutex::new(LazyTransactionRefInner::new(connection_pool))),
+        }
+    }
+
+    pub fn into_maybe_transaction<DB: sqlx::Database>(
+        self,
+    ) -> Option<sqlx::Transaction<'static, DB>> {
+        let m = Arc::try_unwrap(self.inner).unwrap().into_inner();
+        m.maybe_transaction
+            .map(|t| *t.downcast::<sqlx::Transaction<'static, DB>>().unwrap())
+    }
+}
+
+#[derive(Debug)]
+pub struct LazyTransactionRefInner {
+    connection_pool: Box<dyn Any + Send>,
+    maybe_transaction: Option<Box<dyn Any + Send>>,
+}
+
+impl LazyTransactionRefInner {
+    fn new<DB: sqlx::Database>(connection_pool: sqlx::pool::Pool<DB>) -> Self {
+        Self {
+            connection_pool: Box::new(connection_pool),
+            maybe_transaction: None,
+        }
+    }
+}
+
+/// A typed wrapper over the [`TransactionRef`]. It propagates the type
+/// information to [`TransactionGuard`] to safely access typed
+/// [`sqlx::Transaction`] object.
+pub struct TransactionRefT<DB: sqlx::Database> {
+    tr: LazyTransactionRef,
```

Comment:
Not a fan of the `LazyTransaction` name - I feel like laziness should be kept an implementation concern rather than proclaimed in all APIs - in case in future we will want to switch to a model where middlewares open transactions explicitly based on some decorator.

Reply:
Okay